builder/test/simplify.cc

```diff
@@ -414,111 +415,127 @@ TEST(simplify, where_true) {
-  switch (rand() % 2) {
-  case 0: return buffer_min(random_pick(bufs), rand() % max_rank);
-  case 1: return buffer_max(random_pick(bufs), rand() % max_rank);
-  default: return buffer_at(random_pick(bufs));
+  template <typename T>
+  T random_pick(const std::vector<T>& from) {
+    return from[rng_() % from.size()];
+  }
+ 
+public:
+  expr_generator() {
+    vars_ = {x, y, z};
+    bufs_ = {b0, b1};
+
+    buffers_.resize(bufs_.size());
+    for (const var& v : vars_) {
+      var_bounds_[v] = {-max_abs_constant, max_abs_constant};
+    }
   }
-}
 
-expr make_random_expr(int depth);
-
-expr make_random_condition(int depth) {
-  expr a = make_random_expr(depth - 1);
-  expr b = make_random_expr(depth - 1);
-  switch (rand() % 8) {
-  default: return a == b;
-  case 1: return a < b;
-  case 2: return a <= b;
-  case 3: return a != b;
-  case 4: return make_random_condition(depth - 1) && make_random_condition(depth - 1);
-  case 5: return make_random_condition(depth - 1) || make_random_condition(depth - 1);
-  case 6: return !make_random_condition(depth - 1);
+  const symbol_map<interval_expr>& var_bounds() { return var_bounds_; }
+
+  void init_context(eval_context& ctx) {
+    for (const var& v : vars_) {
+      ctx[v] = random_constant();
+    }
+    for (int i = 0; i < static_cast<int>(bufs_.size()); ++i) {
+      ctx[bufs_[i]] = reinterpret_cast<index_t>(&buffers_[i]);
+    }
+    for (auto& b : buffers_) {
+      for (int d = 0; d < max_rank; ++d) {
+        // TODO: Add one to extent because the simplifier assumes buffer_max >= buffer_min. This is not
+        // correct in the case of empty buffers. But do we need to handle empty buffers...?
+        index_t min = random_constant();
+        index_t max = std::max(min + 1, random_constant());
+        b.dim(d).set_bounds(min, max);
+      }
+    }
   }
-}
 
-expr make_random_expr(int depth) {
-  if (depth <= 0) {
-    switch (rand() % 4) {
-    default: return random_pick(vars);
-    case 1: return constant::make(random_constant());
-    case 2: return random_buffer_intrinsic();
+  index_t random_constant(int max = max_abs_constant) { return (rng_() & (2 * max - 1)) - max; }
+
+  expr random_buffer_intrinsic() {
+    switch (rng_() % 2) {
+    case 0: return buffer_min(random_pick(bufs_), static_cast<int>(rng_() % max_rank));
+    case 1: return buffer_max(random_pick(bufs_), static_cast<int>(rng_() % max_rank));
+    default: return buffer_at(random_pick(bufs_));
     }
-  } else {
+  }
+
+  expr make_random_condition(int depth) {
     expr a = make_random_expr(depth - 1);
     expr b = make_random_expr(depth - 1);
-    switch (rand() % 9) {
-    default: return a + b;
-    case 1: return a - b;
-    case 2: return a * b;
-    case 3: return a / b;
-    case 4: return a % b;
-    case 5: return min(a, b);
-    case 6: return max(a, b);
-    case 7: return select(make_random_condition(depth - 1), a, b);
-    case 8: return random_constant();
+    switch (rng_() % 8) {
```

Comment:
Are we intentionally producing more == cases than other cases here?

Reply:
Yes, that should have a comment, added.